toolchain/docs/adding_features.md

```diff
@@ -261,13 +268,38 @@ If the resulting SemIR needs a new instruction:
     };
     ```
 
-Adding an instruction will also require a handler in the Lower step.
+    -   [`sem_ir/inst_kind.h`](/toolchain/sem_ir/inst_kind.h) documents the
+        different options when defining a new instruction, as well as their
+        defaults, see `InstKind::DefinitionInfo`.
+    -   If an instruction doesn't have a `type_id`, it won't produce a value,
+        and so won't be given a label when formatted, and can't be the argument
+        of another instruction.
+    -   If an instruction always produces a type, so it sets
+        `.is_type = InstIsType::Always`, then its `type_id` field will be set to
+        `SemIR::TypeId::TypeType`.
+    -   If an instruction produces a value, but not one that can be used in an
+        expression, you can make a new builtin type for its output. For example,
```

Comment:
I thought namespaces had types because they could show up in expressions as part of name lookup, but never have values. Also, aren't bound methods and witnesses used as part of expressions?

To be sure, I would expect most people shouldn't be adding builtin inst kinds, as noted by the comments there. They're definitely an edge case.

Reply:
I agree it isn't common, but
a) if it does come up, you'd be happy to know how to do it
b) existing instructions you might read about use this feature at times
c) it isn't clear what criteria is being used to decide when to make a new builtin type instead of an instruction without a type id
d) I personally would like the deeper understanding that includes edge cases rather than just recipes for common cases.